lib/link_state.c

```diff
@@ -0,0 +1,911 @@
+	}
+
+	edges_add(&ted->edges, new);
+
+	return new;
+}
+
+void ls_edge_del(struct ls_ted *ted, struct ls_edge *edge)
+{
+	edges_del(&ted->edges, edge);
+	XFREE(MTYPE_LS_DB, edge);
+}
+
+struct ls_edge *ls_find_edge(struct ls_ted *ted, const uint64_t key)
+{
+	struct ls_edge edge = {};
+
+	if (key == 0)
+		return NULL;
+
+	edge.key = key;
+	return edges_find(&ted->edges, &edge);
+}
+
+/**
+ * Link State Subnet Management functions.
+ */
+struct ls_subnet *ls_subnet_add(struct ls_ted *ted,
+				struct ls_prefix *ls_pref)
+{
+	struct ls_subnet *new;
+	struct ls_vertex *vertex;
+
+	if (ls_pref == NULL)
+		return NULL;
+
+	new = XCALLOC(MTYPE_LS_DB, sizeof(struct ls_subnet));
+	new->ls_pref = ls_pref;
+	new->key = ls_pref->pref;
+
+	/* Find Vertex */
+	vertex = ls_find_vertex_by_id(ted, ls_pref->adv);
+	if (vertex != NULL) {
+		new->vertex = vertex;
+		listnode_add(vertex->prefixes, new);
+	}
+
+	subnets_add(&ted->subnets, new);
+
+	return new;
+}
+
+void ls_subnet_del(struct ls_ted *ted, struct ls_subnet *subnet)
+{
+	subnets_del(&ted->subnets, subnet);
+	XFREE(MTYPE_LS_DB, subnet);
+}
+
+struct ls_subnet *ls_find_subnet(struct ls_ted *ted, const struct prefix prefix)
+{
+	struct ls_subnet subnet = {};
+
+	subnet.key = prefix;
+	return subnets_find(&ted->subnets, &subnet);
+}
+
+/**
+ * Link State TED management functions
+ */
+struct ls_ted *ls_ted_new(const uint32_t key, char *name, uint32_t as_number)
+{
+	struct ls_ted *new;
+
+	new = XCALLOC(MTYPE_LS_DB, sizeof(struct ls_ted));
+	if (new == NULL)
+		return new;
+
+	/* Set basic information for this ted */
+	new->key = key;
+	new->as_number = as_number;
+	new->name = name;
+
+	/* Initialize the various RB tree */
+	vertices_init(&new->vertices);
+	edges_init(&new->edges);
+	subnets_init(&new->subnets);
+
+	return new;
+}
+
+void ls_ted_del(struct ls_ted *ted)
+{
+	if (ted == NULL)
+		return;
+
+	/* Release RB Tree */
+	vertices_fini(&ted->vertices);
+	edges_fini(&ted->edges);
+	subnets_fini(&ted->subnets);
+
+	XFREE(MTYPE_LS_DB, ted);
+	ted = NULL;
+}
+
+void ls_connect(struct ls_vertex *vertex, struct ls_edge *edge, bool source)
+{
+	if (vertex == NULL || edge == NULL)
+		return;
+
+	if (source) {
+		listnode_add(vertex->outgoing_edges, edge);
+		edge->source = vertex;
+	} else {
+		listnode_add(vertex->incoming_edges, edge);
+		edge->destination = vertex;
+	}
+}
+
+void ls_disconnect(struct ls_vertex *vertex, struct ls_edge *edge, bool source)
+{
+
+	if (vertex == NULL || edge == NULL)
+		return;
+
+	if (source) {
+		listnode_delete(vertex->outgoing_edges, edge);
+		edge->source = NULL;
+	} else {
+		listnode_delete(vertex->incoming_edges, edge);
+		edge->destination = NULL;
+	}
+}
+
+void ls_connect_vertices(struct ls_vertex *src, struct ls_vertex *dst,
+			 struct ls_edge *edge)
+{
+	if (edge == NULL)
+		return;
+
+	edge->source = src;
+	edge->destination = dst;
+
+	if (src != NULL)
+		listnode_add(src->outgoing_edges, edge);
+
+	if (dst != NULL)
+		listnode_add(dst->incoming_edges, edge);
```

Comment:
Should also check NULL for `src->outgoing_edges`,`src->incoming_edges` ?

Reply:
Add list creation in `ls_vertex_add` function